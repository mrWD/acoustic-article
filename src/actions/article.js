import axios from 'axios';

export const getArticle = async (id) => {
  const { data } = await axios.get(`${process.env.REACT_APP_API_URL}/${id}`);

  const article = Object.entries(data.elements).reduce((res, [key, item]) => ({
    ...res,
    [key]: item.value || item.values,
  }), {});

  return article;
};
