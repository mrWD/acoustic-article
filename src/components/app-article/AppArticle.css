.AppArticle {
  padding-top: 32px;
  padding-bottom: 32px;
  text-align: center;
}

.AppArticle__title {
  margin-top: 0;
  margin-bottom: 0;
  padding-top: 32px;
  padding-bottom: 32px;
  border-top: 2px solid #323232;
  font-size: 24px;
  color: #323232;
}

.AppArticle__img {
  max-width: 100%;
  margin-bottom: 32px;
}

.AppArticle__content-wrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  max-width: 600px;
  margin-left: auto;
  margin-right: auto;
}

.AppArticle__info {
  margin-bottom: 16px;
  font-size: 16px;
  color: #323232;
}

.AppArticle__text {
  flex-basis: 100%;
  width: 100%;
  margin-top: 0;
  margin-bottom: 0;
  font-size: 18px;
  text-indent: 24px;
  line-height: 1.5;
  text-align: left;
  color: #323232;
}

.AppArticle__text p {
  margin-top: 0;
  margin-bottom: 0;
}

.AppArticle__text p:not(:last-child) {
  margin-bottom: 16px;
}
