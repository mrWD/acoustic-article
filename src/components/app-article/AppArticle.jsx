import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import moment from 'moment';

import { getArticle } from '../../actions/article';
import { getMsgByError } from '../../helpers/getMsgByError';

import AppPicture from '../app-picture/AppPicture';
import AppLoader from '../app-loader/AppLoader';
import AppErrors from '../app-errors/AppErrors';

import './AppArticle.css';

const AppArticle = () => {
  const { id } = useParams();
  const [isLoading, setLoading] = useState(true);
  const [errors, setErrors] = useState('');
  const [article, setArticle] = useState({});

  useEffect(() => {
    (async () => {
      try {
        const article = await getArticle(id);

        setArticle(article);
      } catch (error) {
        const errorMsg = getMsgByError(error.response.status);

        setErrors(errorMsg);
      } finally {
        setLoading(false);
      }
    })();
  }, [id]);

  if (isLoading) {
    return <AppLoader />;
  }

  if (errors) {
    return <AppErrors errors={errors} />;
  }

  const body = article.body?.join('');
  const date = moment(article.date).format('DD.MM.YYYY');

  return (
    <article className="AppArticle">
      <h1 className="AppArticle__title">
        {article.heading}
      </h1>

      <AppPicture
        className="AppArticle__img"
        leadImage={article.mainImage?.leadImage}
        altText={article.mainImage?.leadImageCaption.value}
      />
  
      <div className="AppArticle__content-wrapper">
        <span className="AppArticle__info">
          {article.author}
        </span>
  
        <span className="AppArticle__info">
          {date}
        </span>

        <div
          className="AppArticle__text"
          dangerouslySetInnerHTML={{ __html: body }}
        ></div>
      </div>
    </article>
  );
};

export default AppArticle;
