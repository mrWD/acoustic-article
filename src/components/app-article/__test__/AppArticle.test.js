import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, configure } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { getArticle } from '../../../actions/article';

import AppArticle from '../AppArticle';

configure({ adapter: new Adapter() });

jest.mock('../../../actions/article.js', () => ({
  getArticle: jest.fn().mockResolvedValue({
    heading: 'Article title',
    author: 'Article Author',
    body: ['<p>Article text</p>'],
    date: '2019-12-13T17:00:00Z',
    mainImage: {
      leadImage: {
        url: '/img-url.jpeg',
        renditions: {
          default: {
            width: 1600,
            url: '/img-url.jpeg',
          },
        },
      },
      leadImageCaption: {
        value: 'Image caption.',
      },
    },
  }),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    push: jest.fn(),
  }),
}));

describe('AppArticle', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = null;
  });

  test('renders without crashing', async () => {
    await act(async () => {
      wrapper = await shallow(<AppArticle />);
    });

    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  test('renders article content', async () => {
    await act(async () => {
      wrapper = await mount(<AppArticle />);
    });

    wrapper.update();


    expect(wrapper).toMatchSnapshot();
  });

  test('returns errors', async () => {
    getArticle.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    await act(async () => {
      wrapper = await mount(<AppArticle />);
    });

    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});
