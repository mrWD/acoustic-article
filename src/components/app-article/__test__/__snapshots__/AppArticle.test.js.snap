// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppArticle renders article content 1`] = `
<AppArticle>
  <article
    className="AppArticle"
  >
    <h1
      className="AppArticle__title"
    >
      Article title
    </h1>
    <AppPicture
      altText="Image caption."
      className="AppArticle__img"
      leadImage={
        Object {
          "renditions": Object {
            "default": Object {
              "url": "/img-url.jpeg",
              "width": 1600,
            },
          },
          "url": "/img-url.jpeg",
        }
      }
    >
      <picture>
        <source
          key="/img-url.jpeg"
          media="(max-width: 1600px)"
          srcSet="https://content-eu-4.content-cms.com/img-url.jpeg"
        />
        <img
          alt="Image caption."
          className="AppArticle__img"
          src="https://content-eu-4.content-cms.com/img-url.jpeg"
        />
      </picture>
    </AppPicture>
    <div
      className="AppArticle__content-wrapper"
    >
      <span
        className="AppArticle__info"
      >
        Article Author
      </span>
      <span
        className="AppArticle__info"
      >
        13.12.2019
      </span>
      <div
        className="AppArticle__text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>Article text</p>",
          }
        }
      />
    </div>
  </article>
</AppArticle>
`;

exports[`AppArticle renders without crashing 1`] = `<AppLoader />`;

exports[`AppArticle returns errors 1`] = `
<AppArticle>
  <AppErrors
    errors="Sorry, but the article does not exist! Try another article ID."
  >
    <p
      className="AppErrors"
    >
      Sorry, but the article does not exist! Try another article ID.
    </p>
  </AppErrors>
</AppArticle>
`;
